src/main/java/net/neoforged/neoforge/transfer/transaction/snapshots/IndexedIntSnapshot.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.transaction.snapshots;
+
+import java.util.ArrayList;
+import java.util.List;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.neoforged.neoforge.transfer.transaction.SnapshotJournal;
+import net.neoforged.neoforge.transfer.transaction.TransactionContext;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A snapshot journal that can keep track of {@code int} values in an indexable structure. Also takes in a SnapshotJournal
+ * that can be updated when an index gets written to. An example is using a {@link GroupedSnapshotJournal} to handle
+ * the scenario of doing a single commit for something like {@link BlockEntity#setChanged()}.
+ */
+public class IndexedIntSnapshot extends SnapshotJournal<Integer> {
+    /**
+     * Apply the value for snapshotting at the specified index. This value should be the last valid value from the {@link Snapshot}
+     * during the transaction chain.
+     */
+    @FunctionalInterface
+    public interface Revert {
+        void set(int index, int value);
+    }
+
+    /**
+     * Gets the current value for snapshotting. Given some index, what would the value be if we were to need to revert.
+     */
+    @FunctionalInterface
+    public interface Snapshot {
+        Integer get(int index);
+    }
+
+    private final int index;
+    private final Revert setter;
+    private final Snapshot getter;
+    //One shared setChanged journal shared across all indices
+    @Nullable
+    private final SnapshotJournal<?> onChangedSnapshot;
+
+    /**
+     * Returns a snapshot handling a specific index of a container for integers.
+     * 
+     * @param setter   The reversion method to apply the snapshot value should a transaction fail.
```

Comment:
This should be called "reverter", or some such. Or, in transaction terms, something like "rollback"